File: client/src/lib/blaze.ts
// Client's half of blaze rendering (see pipeline/lib/blaze.py for the other
// half - decoding raw ArcGIS codes into a normalized `blaze_color` string,
// done during pipeline export, not here). This module's only job: map an
// already-normalized blaze_color to a paint color, with a defensive
// fallback + warning for anything unexpected - belt-and-suspenders, since a
// rendering layer shouldn't blindly trust upstream data. See WIREFRAMES.md's
// "Trail line rendering — blazes" section for the exact hex values, which
// are load-bearing, not decorative.

/**
 * The grey a trail line takes when its blaze is not known.
 *
 * Exported since #598. The corridor view spends the same grey on the miles
 * ATC's centerline records no maintaining club for - the same sentence, "we do
 * not know this", about a different question - and sharing the constant is what
 * keeps one sentence one colour. WIREFRAMES.md line 325 pins the value.
 */
export const NEUTRAL_BLAZE_COLOR = '#8a8271'

const NEUTRAL_FALLBACK = NEUTRAL_BLAZE_COLOR

/**
 * THE PALETTE IS CLOSED, AND THIS IS WHERE SPRAWL STOPS (#782).
 *
 * The maintainer's decision, verbatim (features/NEARBY_TRAILS.md §4,
 * 2026-08-18): *"we will need to bring in more colors for the blazes. Long
 * [Path] is indeed aqua. Some way to stop sprawl is needed, but the color of
 * the trail blazes should be the color on the map."*
 *
 * Both halves of that are load-bearing. The paint on the tree is what a hiker
 * navigates by, so a map that flattens it is lying about the ground — and a
 * palette that grows every time a new source arrives ends up with forty hues
 * nobody can tell apart, which lies about it differently. So this object is
 * the ENTIRE set of hues this map will ever paint, and it grows by review
 * rather than by data arrival.
 *
 * **A new member needs three things, and the third is the one that gets
 * skipped:**
 *
 *   1. a real trail wearing it — measured on a real source, not anticipated;
 *   2. a hex that stays separable from its nearest palette neighbour and
 *      readable on both the day and the night sheet, with the numbers written
 *      down here rather than asserted;
 *   3. no change to the red-light collapse. That one is free by construction
 *      and worth knowing why: `map/style.ts`'s `blazeLineColor` replaces this
 *      whole expression with `RED_LIGHT_BLAZE_COLOR` under red light, so a
 *      member added here cannot reach that mode at all. Blaze identity moves
 *      to the tapped trail's sheet there, honestly, rather than pretending to
 *      survive on the line.
 *
 * **The bar, measured against the palette itself rather than picked** (all
 * figures 2026-08-22, CIE76 ΔE in Lab and WCAG contrast ratio):
 *
 *   - **Separation ≥ 24.178.** That is Blue/Purple, the closest pair already
 *     shipping. A new hue at least as separable as the worst pair a hiker
 *     already reads is not a regression; a tighter one is.
 *   - **Day contrast ≥ 2.076** against `#ffffff`, the field sheet's paper —
 *     Yellow's, the lowest of any hue that is not White. (White itself is
 *     1.02 and stays: the AT's centerline is white paint, and its width and
 *     casing are what carry it.)
 *   - **Night contrast ≥ 2.66** against `#0c1410`, night_hike's ink —
 *     Purple's, the lowest shipping.
 *
 * These are arithmetic, and arithmetic is not the whole of legibility:
 * `@unvalidated` — nothing here has been read on a real phone in real
 * sunlight, which is **#105 — Outdoor usability pass** and the thing that
 * would actually settle a hex. What would change a number is that pass
 * reporting a hue that washes out, not a nicer figure in this comment.
 */
const BLAZE_COLORS: Record<string, string> = {
  White: '#fffdf7',
  Blue: '#1f5fa8',
  Yellow: '#dcae1b',
  Orange: '#d2721c',
  Red: '#b2321f',
  Green: '#2f7a44',
  Purple: '#6a4a8f',
  // FIRST ADMISSION UNDER THE RULE ABOVE (#782), and the Long Path forces it:
  // 107 Aqua + 28 Teal rows on OPRHP's own Long Path segments, agreeing with
  // NYNJTC's separate layer (measured 2026-08-18, #771). Aqua is real paint on
  // real trees, not a source's idiosyncratic spelling.
  //
  // Measured 2026-08-22 against the three bars above: nearest neighbour is the
  // neutral grey at ΔE 36.6 (Green 37.3, Blue 49.9), against a 24.178 bar;
  // day contrast 3.90 against a 2.076 bar; night contrast 4.80 against 2.66.
  // Chosen over #00a0a8, which separates slightly better (ΔE 39.9) and reads
  // worse where it matters more — 3.18 on the day sheet, which is the one a
  // hiker holds in the sun.
  Aqua: '#0d8f96',
  // All three are real values from the pipeline (see pipeline/lib/blaze.py) -
  // "None" means confirmed unblazed, "Other" is a real domain value with no
  // dedicated paint style, and "Unknown" is what its NEUTRAL_FALLBACK emits
  // for every undecodable blaze, BY CONTRACT. Warning on a value our own
  // upstream guarantees will occur was noise dressed as vigilance (#257);
  // the warning below is kept for genuinely novel values.
  None: NEUTRAL_FALLBACK,
  Other: NEUTRAL_FALLBACK,
  Unknown: NEUTRAL_FALLBACK,
}

/**
 * The values that mean "no hue to draw" rather than naming one.
 *
 * All three are real pipeline output, not error states: "None" is a confirmed
 * unblazed trail, "Other" is a real domain value with no dedicated paint, and
 * "Unknown" is what `lib/blaze.py` emits by contract for every value it could
 * not decode.
 */
export const NEUTRAL_BLAZE_MEMBERS = ['None', 'Other', 'Unknown'] as const

/**
 * Every hue the palette admits, in declaration order.
 *
 * Derived rather than listed, so that admitting a member reaches everything
 * that enumerates the palette — the legend's rows, the style tests, and the
 * governance check in blaze.test.ts — without anybody remembering to add it
 * in a second place. #782 makes that a requirement rather than a nicety:
 * *"the legend's blaze rows pick up new members automatically or this issue
 * is not done."*
 */
export const BLAZE_PALETTE_MEMBERS = Object.keys(BLAZE_COLORS).filter(
  (name) => !(NEUTRAL_BLAZE_MEMBERS as readonly string[]).includes(name),
)

/** Every blaze_color string this module answers for, hues and neutrals alike. */
export const BLAZE_MEMBERS = Object.keys(BLAZE_COLORS)

/**
 * One palette member's hex, by name.
 *
 * NO PRODUCTION CALLER SINCE 2026-08-25, and that is worth knowing before
 * reading this as live rendering code. The map has never painted through it -
 * `BLAZE_MATCH_EXPRESSION` below is what `map/style.ts` hands MapLibre - and
 * its one caller was the legend's blaze swatch, removed with those rows
 * (chrome/Legend.tsx's header has the decision).
 *
 * It stays because `BLAZE_COLORS` is not exported, so this is the only way to
 * read a member's hex, and `lib/blazeGovernance.test.ts` reads every member
 * through it to hold #782's admission bar - the ΔE separation and the day and
 * night contrast ratios. Deleting it would mean opening the table up to
 * exactly the sprawl that issue closed off.
 */
export function blazePaintColor(blazeColor: string): string {
  if (typeof blazeColor === 'string' && blazeColor in BLAZE_COLORS) {
    return BLAZE_COLORS[blazeColor]
  }
  console.warn(`Unrecognized blaze_color "${blazeColor}" - rendering as neutral grey.`)
  return NEUTRAL_FALLBACK
}

// One MapLibre `match` expression for `line-color`, not per-layer hardcoding
// - centerline, side_trails, and anything imported later all share this one
// rendering rule (TRAIL_BLAZE_COLORS.md).
export const BLAZE_MATCH_EXPRESSION = [
  'match',
  ['get', 'blaze_color'],
  ...Object.entries(BLAZE_COLORS)
    .filter(([color]) => color !== 'None' && color !== 'Other')
    .flatMap(([color, hex]) => [color, hex]),
  'None',
  NEUTRAL_FALLBACK,
  'Other',
  NEUTRAL_FALLBACK,
  NEUTRAL_FALLBACK, // fallback for anything else (Unknown, Gold, etc.)
] as const

/**
 * How each `blaze_color` is NAMED, wherever a surface says one out loud.
 *
 * The pipeline's contract above is what makes this more than a capitalisation
 * helper: "None" is a CONFIRMED unblazed trail while "Unknown" is a value
 * that failed to decode, and the two are different claims. WIREFRAMES.md §3
 * requires saying plainly when a blaze is unknown, and "Unblazed" in that
 * slot would be a confident statement nobody made.
 *
 * Lifted here from lib/lineDetail.ts when the turn card (#1041) became the
 * second surface naming a blaze. It is the same rule in both places by
 * construction now: a hiker who reads "Blaze not recorded" on a tapped line
 * and "Unblazed" at the junction it leads to has been told two different
 * things about one piece of tread.
 */
export function blazeLabel(blazeColor: string | null): string {
  if (blazeColor === null || blazeColor === 'Unknown') return 'Blaze not recorded'
  if (blazeColor === 'None') return 'Unblazed'
  if (blazeColor === 'Other') return 'Other blaze'
  return `${blazeColor} blaze`
}

File: pipeline/lib/blaze.py
"""Normalize raw trail-line blaze-color values into one `blaze_color`
attribute, per features/TRAIL_BLAZE_COLORS.md.

Pure - no I/O, no logging. Deciding what "decoded" means is this module's
job; deciding whether to warn about a failed decode is the caller's (during
export, where the raw feature/source context is available to put in the
message).
"""

NEUTRAL_FALLBACK = "Unknown"

# The palette the client will actually paint, mirrored here so that a mapping
# table cannot name a member that does not exist (#782).
#
# WHY THIS IS A SECOND COPY, AND WHAT STOPS IT DRIFTING
#
# The one that renders is `client/src/lib/blaze.ts`'s `BLAZE_COLORS`, and this
# module cannot import it - different language, different package. So this is
# the same shape as the POI id resolver (#831): one list per runtime, held to
# the other by a contract test rather than by anybody remembering.
# `tests/test_blaze_palette_contract.py` reads the TypeScript and fails if the
# two disagree, and pipeline-tests.yml's scope list carries that file so
# editing the palette runs this suite too.
#
# The neutrals are here for the same reason a mapping table might legitimately
# name one: "this source's blank string means confirmed-unblazed" is a real
# reviewed decision, distinct from "we could not decode it".
PALETTE = (
    "White",
    "Blue",
    "Yellow",
    "Orange",
    "Red",
    "Green",
    "Purple",
    "Aqua",
)

NEUTRAL_MEMBERS = ("None", "Other", "Unknown")


class UnknownPaint(ValueError):
    """A mapping table names a palette member the client cannot paint.

    Raised rather than warned, and that is the point: a warning would let a
    release ship trails coloured by a member that renders as neutral grey
    everywhere, which looks like missing data rather than like a typo in a
    reviewed file. This is a file a person edited; the failure belongs at the
    edit, not on a phone.
    """


def load_blaze_mapping(path=None) -> dict:
    """The reviewed per-source tables, or an empty mapping.

    The one impure function in this module, and it is here rather than at the
    call site so that "which file holds the judgement" has one answer. Absent
    file reads as no tables, which is the honest state of a checkout that has
    not fetched anything: every value unmapped, every one warned about.
    """
    import json
    from pathlib import Path

    path = path or Path(__file__).resolve().parent.parent / "reference" / "blaze_mapping.json"
    if not path.exists():
        return {}
    return json.loads(path.read_text(encoding="utf-8")).get("sources", {})


def map_source_blaze(raw_value, table: dict | None) -> tuple[str, str]:
    """Resolve one source's raw blaze string to (palette member, disposition).

    The other half of `normalize_blaze_color` and deliberately separate from
    it. That function decodes an ArcGIS coded domain - a mechanical step, the
    same for everyone. This one applies JUDGEMENT recorded in
    `reference/blaze_mapping.json`: that OPRHP's "Teal" is the same paint a
    hiker sees as aqua, that its "Lime" might be Green and nobody has checked.
    Two different kinds of decision, kept apart so a reviewer reads the second
    without wading through the first.

    Three dispositions, and the middle one is why this returns a word rather
    than a bool:

      - `"mapped"` - a reviewed row named a palette member.
      - `"deferred"` - a value this project has SEEN and decided not to paint
        yet, with a reason in the file. Renders neutral, same as unmapped, and
        it is not the same event: one is a decision, the other is an
        oversight. Collapsing them is how an oversight hides inside a docket.
      - `"unmapped"` - nobody has looked at this value. The loud one.

    A missing table is not an error: a source with no reviewed mapping has
    every value unmapped, which is exactly what the first release of a new
    source should say out loud.
    """
    if raw_value is None:
        return NEUTRAL_FALLBACK, "unmapped"
    table = table or {}
    mapped = (table.get("mapped") or {}).get(raw_value)
    if mapped is not None:
        if mapped not in PALETTE and mapped not in NEUTRAL_MEMBERS:
            raise UnknownPaint(
                f"blaze mapping names {mapped!r}, which is not a palette member - "
                f"admit it in client/src/lib/blaze.ts first, or map to one of {PALETTE}"
            )
        return mapped, "mapped"
    if raw_value in (table.get("deferred") or {}):
        return NEUTRAL_FALLBACK, "deferred"
    return NEUTRAL_FALLBACK, "unmapped"


def normalize_blaze_color(raw_value, coded_domain: dict[int, str] | None, source_default: str | None = None) -> tuple[str, bool]:
    """Resolve one feature's raw blaze value to (blaze_color, decoded).

    - `raw_value` missing (None): a source with a flat per-source default
      (e.g. centerline, uniformly white with no per-feature field - see
      `blaze_default` in sources.json) resolves to that default, decoded=True.
      Without a default, this is a true non-decode: (Unknown, False).
    - `raw_value` is a real key in `coded_domain`: decode it. This includes
      code 0 -> "None" and code 9 -> "Other" (side_trails' real domain) -
      both are successful decodes, not fallbacks.
    - Anything else (an unmapped literal like "Unknown"/"Gold", an
      out-of-range code, or no domain and no default to fall back on): the
      neutral (Unknown, False) fallback.
    """
    if raw_value is None:
        if source_default is not None:
            return source_default, True
        return NEUTRAL_FALLBACK, False
    if coded_domain is not None and raw_value in coded_domain:
        return coded_domain[raw_value], True
    return NEUTRAL_FALLBACK, False
